src/ucp/proto/proto_debug.c

```diff
@@ -198,6 +205,77 @@ static int ucp_proto_debug_is_info_enabled(ucp_context_h context,
 }
 
+static ucs_status_t
+ucp_proto_debug_add_info_row(const ucp_proto_debug_range_info_t *range_info,
+                             ucs_array_t(ucp_proto_info_table) *table,
+                             int *col_width)
+{
+    const ucp_proto_query_attr_t *proto_attr = &range_info->proto_attr;
+    ucp_proto_info_row_t *row_elem;
+
+    row_elem = ucs_array_append(ucp_proto_info_table, table,
+                                return UCS_ERR_IO_ERROR);
+
+    ucs_snprintf_safe(row_elem->desc, sizeof(row_elem->desc), "%s%s",
+                      proto_attr->is_estimation ? "(?) " : "",
+                      proto_attr->desc);
+    ucs_strncpy_safe(row_elem->config, proto_attr->config,
+                     sizeof(row_elem->config));
+
+    ucs_memunits_range_str(range_info->range_start, range_info->range_end,
+                           row_elem->range_str, sizeof(row_elem->range_str));
+
+    col_width[0] = ucs_max(col_width[0], strlen(row_elem->range_str));
+    col_width[1] = ucs_max(col_width[1], strlen(row_elem->desc));
+    col_width[2] = ucs_max(col_width[2], strlen(row_elem->config));
+
+    return UCS_OK;
+}
+
+static int
+ucp_proto_debug_get_range_info(ucp_worker_h worker,
+                               const ucp_proto_select_elem_t *select_elem,
+                               ucp_proto_debug_range_info_t *range_info)
+{
+    const ucp_proto_perf_range_t *range;
+
+    range_info->range_start = range_info->range_end + 1;
+    range = ucp_proto_perf_range_search(select_elem, range_info->range_start);
+    range_info->range_end = range->max_length;
+
+    return ucp_proto_select_elem_query(worker, select_elem,
+                                       range_info->range_start,
+                                       &range_info->proto_attr);
+}
+
+static void
+ucp_proto_debug_copy_range_info(const ucp_proto_debug_range_info_t *src,
+                                ucp_proto_debug_range_info_t *dst)
+{
+    const ucp_proto_query_attr_t *proto_attr_src = &src->proto_attr;
+    ucp_proto_query_attr_t *proto_attr_dst       = &dst->proto_attr;
+
+    ucs_strncpy_safe(proto_attr_dst->desc, proto_attr_src->desc,
+                     sizeof(proto_attr_dst->desc));
+    ucs_strncpy_safe(proto_attr_dst->config, proto_attr_src->config,
+                     sizeof(proto_attr_dst->config));
+    proto_attr_dst->is_estimation = proto_attr_src->is_estimation;
+    dst->range_start              = src->range_start;
+    dst->range_end                = src->range_end;
+}
+
+static int
+ucp_proto_debug_cmp_range_info(const ucp_proto_debug_range_info_t *range_info_1,
+                               const ucp_proto_debug_range_info_t *range_info_2)
+{
+    const ucp_proto_query_attr_t *proto_attr_1 = &range_info_1->proto_attr;
+    const ucp_proto_query_attr_t *proto_attr_2 = &range_info_2->proto_attr;
+
+    return proto_attr_1->is_estimation == proto_attr_2->is_estimation &&
+           strcmp(proto_attr_1->config, proto_attr_2->config) == 0 &&
```

Comment:
do you assume that `lane_map` is also equal if configs are equal? If yes, can we add some assert ?